### PR TITLE
Add repo cloning to check-diff crate

### PR DESCRIPTION
**Pull request by benluiwj**
TLDR: Added repo cloning to check-diff. Included sample execution as screenshot.

This PR adds git functionalities to the `check-diff` crate. The current script has a `init_submodules` function. However, it seems that the repos used does not require initialising of submodules. Hence, this function was not converted to Rust. 
Sample execution attached. Parameters to the program is currently not in used.
Screenshot 2024-06-06 at 9 49 18 PM
The screenshot shows the following things:
- runs the script to clone `rustfmt` into a `tmp` directory created
- show that `tmp` contains the entire `rustfmt` repo

### CODE REVIEW

File: check_diff/src/main.rs [hunk 1]
```diff
@@ -15,11 +20,36 @@ struct CliInputs {
...
 }
 
-fn main() {
-    let args = CliInputs::parse();
+/// Clone a git repository and cd into it.
+///
+/// Parameters:
+/// url: git clone url
+/// dest: directory where the repo should be cloned
+fn clone_git_repo(url: &str, dest: &str) {
```
Comment (ytmimi) [inline check_diff/src/main.rs:28]: Better to make `dest` a `PathBuf`, `&Path`, or `AsRef<Path>` so we can have a stronger typed API instead of relying on strings. You may even consider adding something like the [`url`](https://docs.rs/url/latest/url/index.html) crate to strongly type the repo URL, though maybe that's overkill.

Also, why not have `clone_git_repo` return `Result<()>` to better communicate to callers if the repo was cloned successfully. Additionally, I don't think `clone_git_repo` should be responsible for both cloning the repo and `cd`-ing into the directory.

File: check_diff/src/main.rs [hunk 2]
```diff
@@ -15,11 +20,36 @@ struct CliInputs {
...
-        args.remote_repo_url, args.feature_branch, args.commit_hash, args.rustfmt_config
+        "Current directory: {}",
+        env::current_dir().unwrap().display()
     );
 }
+
+fn main() {
+    let _args = CliInputs::parse();
+    let sample_repo = "https://github.com/rust-lang/rustfmt.git";
+    clone_git_repo(sample_repo, "./tmp/");
```
Comment (ytmimi) [inline check_diff/src/main.rs:54]: What do you think of writing a unit test for `clone_git_repo`? That might be a better more reliable way to ensure that the functionality is working as you intend.

File: check_diff/src/main.rs [hunk 3]
```diff
@@ -15,11 +20,36 @@ struct CliInputs {
...
+        .output()
+        .expect("failed to clone repository");
+    // if the git command does not return successfully,
+    // cd command will not work as well. So fail fast.
+    if !git_cmd.status.success() {
+        io::stdout().write_all(&git_cmd.stdout).unwrap();
+        io::stderr().write_all(&git_cmd.stderr).unwrap();
+        return;
+    }
+    println!("Successfully clone repository. Entering repository");
```
Comment (ytmimi) [inline check_diff/src/main.rs:41]: Not something we have to change right now, but I'd love to replace these `println!` logs with [`tracing`](https://docs.rs/tracing/latest/tracing/) or the [`log`](https://docs.rs/log/latest/log/) crate. It probably doesn't make much of a difference, though we already use `tracing` within rustfmt.

File: check_diff/src/main.rs [hunk 4]
```diff
@@ -15,11 +20,36 @@ struct CliInputs {
...
 
-fn main() {
-    let args = CliInputs::parse();
+/// Clone a git repository and cd into it.
+///
+/// Parameters:
+/// url: git clone url
+/// dest: directory where the repo should be cloned
+fn clone_git_repo(url: &str, dest: &str) {
+    env::set_var("GIT_TERMINAL_PROMPT", "0");
```
Comment (ytmimi) [inline check_diff/src/main.rs:29]: [Just a heads up `env::set_var` is going to be made unstable in rust 2024](https://github.com/rust-lang/rust/pull/124636). It's probably better to set the `GIT_TERMINAL_PROMT` in this case using the [`Command::env`](https://doc.rust-lang.org/std/process/struct.Command.html#method.env) builder method.

**Review by ytmimi — CHANGES_REQUESTED** [review]
Great first pass, though I think we can make some changes to take better advantage of the type system, and keep functions more focused on a single task. There are also a few other suggestions that I've made. Take a look and let me know if you've got any questions on the feedback.

File: check_diff/tests/git.rs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    );
+    let path = Path::new(TEMP_DIR_PATH);
+    println!("{}", path.exists());
+    if path.exists() {
+        let _ = fs::remove_dir_all(path);
+    }
+}
+
+#[test]
+fn clone_repo_test() {
```
Comment (benluiwj) [inline check_diff/tests/git.rs:25]: @ytmimi for this test fails when i run `cargo test`. however, when running it on its own, it works fine. For some reason after `clone_git_repo` is called, the current directory becomes `./tmp/`. This is quite unexpected since i did not change directory in `clone_git_repo`.

**Comment by benluiwj** [comment]
@ytmimi For the tests, I am currently performing cleanups before and after the test is done. Im not sure whether there is a better way to do so. Let me know what do you think about this?

File: check_diff/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        let errors = GitErrors {
+            stderr: git_cmd.stderr,
+            stdout: git_cmd.stdout,
+        };
+        return Err(errors);
+    } else {
+        event!(Level::INFO, "Successfully clone repository.");
+        return Ok(());
+    }
+}
```
Comment (ytmimi) [inline check_diff/src/lib.rs:44]: If we're returning a `Result` I don't think we should call methods like `expect` or `unwrap` that could panic. Probably better to extend `GitError` with a `IO` variant.

Also, I don't think we should emit anything to `stdout` or `stderr` here. It's better not to emit anything from the library code and let the binary that we're builder handle all that.

Oh and you can simplify the `event!(..)` calls to just be `info!(..)`

```rust
use tracing::info;

enum GitError {
    FailedClone { stdout: Vec<u8>, stderr: Vec<u8> },
    IO(std::io::Error),
}

/// Clone a git repository
///
/// Parameters:
/// url: git clone url
/// dest: directory where the repo should be cloned
pub fn clone_git_repo(url: &str, dest: &Path) -> Result<(), GitError> {
    let git_cmd = Command::new("git")
        .env("GIT_TERMINAL_PROMPT", "0")
        .args([
            "clone",
            "--quiet",
            url,
            "--depth",
            "1",
            dest.to_str().unwrap(),
        ])
        .output()?;

    // if the git command does not return successfully,
    // any command on the repo will fail. So fail fast.
    if !git_cmd.status.success() {
        let error = GitError::CloneFailed { stdout: git_cmd.stdout, stderr: git_cmd.stderr };
        return Err(error);
    }

    info!("Successfully clone repository.");
    return Ok(());
}
```

File: check_diff/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+pub fn change_directory_to_path(dest: &Path) {
+    let dest_path = Path::new(&dest);
+    let _ = env::set_current_dir(&dest_path).is_ok();
+    event!(
+        Level::INFO,
+        "Current directory: {}",
+        env::current_dir().unwrap().display()
+    );
+}
```
Comment (ytmimi) [inline check_diff/src/lib.rs:54]: Similar change to use `info!` here. Also, given that `env::set_current_dir` has the potential to fail, we probably want to bubble that information up to the caller by returning the `std::io::Result<()>`.

```suggestion
pub fn change_directory_to_path(dest: &Path) -> std::io::Result<()> {
    let dest_path = Path::new(&dest);
    env::set_current_dir(&dest_path)?;
    info!("Current directory: {}", env::current_dir().unwrap().display());
}
```

**Review by ytmimi — CHANGES_REQUESTED** [review]
Great to see that you've added some test cases to cover the functionality that you're adding. I left a few more comments about returning `Results` from functions instead of panicing and also I think the `tempfile` crate would be a nice addition to the test suite so that we don't need to roll our own temporary file and directory cleanup.

**Comment by benluiwj** [comment]
Changes made from last review:
- Added `tempfile` crate for unit tests
- Polished `Result` of git (clone) and unix (cd) functionalities

File: check_diff/tests/bash_commands.rs [hunk 8]
```diff
@@ -0,0 +1,34 @@
+use check_diff::change_directory_to_path;
+use std::env;
+use std::ffi::OsStr;
+use tempfile::Builder;
+use tracing::error;
+
+#[test]
+fn cd_test() {
+    // create directory in current directory
+    let dir = Builder::new().tempdir_in("");
```
Comment (ytmimi) [inline check_diff/tests/bash_commands.rs:10]: I don't think we need to use the builder here. Probably simpler to use the [tempdir](https://docs.rs/tempfile/latest/tempfile/fn.tempdir.html) function and just `unwrap` since we're in a test case. If we fail to create the temporary directory then there's not much more we can do.

File: check_diff/tests/bash_commands.rs [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+                env::current_dir().unwrap().file_name(),
+                Some(OsStr::new("check_diff"))
+            );
+            match d.close() {
+                Ok(_) => {}
+                Err(e) => {
+                    error!("Error from closing: {}", e);
+                    assert_eq!(1, 2);
+                }
+            }
```
Comment (ytmimi) [inline check_diff/tests/bash_commands.rs:27]: Can you explain why we need to call `close` here.

File: check_diff/tests/bash_commands.rs [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+#[test]
+fn cd_test() {
+    // create directory in current directory
+    let dir = Builder::new().tempdir_in("");
+
+    match dir {
+        Ok(d) => {
+            let dest_path = d.path();
+            let _ = change_directory_to_path(dest_path);
```
Comment (ytmimi) [inline check_diff/tests/bash_commands.rs:15]: Probably simpler to just `unwrap` in the test instead of silently ignoring the returned `Result`.

File: check_diff/tests/bash_commands.rs [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+            assert_eq!(env::current_dir().is_ok(), true);
```
Comment (ytmimi) [inline check_diff/tests/bash_commands.rs:16]: I feel like it would be a more direct test if you asserted that the current working directory is the same as the `dest_path`

File: check_diff/tests/bash_commands.rs [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+            assert_ne!(
+                env::current_dir().unwrap().file_name(),
+                Some(OsStr::new("check_diff"))
+            );
```
Comment (ytmimi) [inline check_diff/tests/bash_commands.rs:20]: This is the code you were referring to in https://github.com/rust-lang/rustfmt/pull/6187#issuecomment-2196426926, right? My feeling is that this isn't a necessary check and if we assert that we've correctly changed the working directory to the tempdir then I don't think this is necessary.

File: check_diff/tests/git.rs [hunk 13]
```diff
@@ -0,0 +1,34 @@
+use check_diff::clone_git_repo;
+
+use std::fs;
+use tempfile::Builder;
+use tracing::error;
+
+#[test]
+fn clone_repo_test() {
+    let dir = Builder::new().tempdir_in("");
```
Comment (ytmimi) [inline check_diff/tests/git.rs:9]: Similar comment as before. It's probably simpler to use the [tempdir](https://docs.rs/tempfile/latest/tempfile/fn.tempdir.html) function and just `unwrap` if we fail to create a tempdir.

File: check_diff/tests/git.rs [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+        Ok(d) => {
+            let sample_repo = "https://github.com/rust-lang/rustfmt.git";
+            let dest_path = d.path();
+            let result = clone_git_repo(sample_repo, dest_path);
+            assert_eq!(result.is_ok(), true);
+            let directory = fs::read_dir(dest_path);
+            // check whether we can read this directory
+            assert_eq!(directory.is_err(), false);
+            // check that the directory is non-empty
+            assert_eq!(directory.iter().next().is_none(), false);
```
Comment (ytmimi) [inline check_diff/tests/git.rs:20]: Since we've cloned a git repo into this repository we might want to double check that a `.git` directory exists with the `tempdir`. That might be a more direct assertion. Checking that we're able to reach the directly could be useful, but my assumption would be that if we're able to write to the tempdir then we'd be able to read from it as well.

File: check_diff/tests/git.rs [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+            match d.close() {
+                Ok(_) => {}
+                Err(e) => {
+                    error!("Error from closing: {}", e);
+                    assert_eq!(1, 2);
+                }
+            }
+        }
```
Comment (ytmimi) [inline check_diff/tests/git.rs:28]: Similar question as before. I'm not sure if we need to manually close the directory since it should be closed and cleaned up at the end of the test already.

**Review by ytmimi — CHANGES_REQUESTED** [review]
I left some additional comments on the `cd_test` and `clone_repo_test` functions. When testing I think it's totally fine to call `unwrap`, and I think doing so would simplify the test cases. I also left some suggestions for more direct assertions. Overall I think the addition of the `tempfile` crate helped improve the test cases.

File: check_diff/tests/bash_commands.rs (hunk 10 above)
Comment (benluiwj) [reply check_diff/tests/bash_commands.rs:15]: the `Result` here is actually ignored because we are actually changing the `env` variable.

**Comment by benluiwj** [comment]
Updated most of the changes that were requested as I agreed with them. The use of `Builder` is actually to force `tempfile` to create directory in the current crate directory **instead of** at `std::env::temp_dir()`. I suspect that if the testcase were to fail, the directory will not be deleted. Thus, having the temporary directory created in the same crate allows the user to manually delete the folder if that happens.

File: check_diff/tests/git.rs [hunk 16]
```diff
@@ -1,33 +1,25 @@
...
-        }
-        Err(e) => {
-            error!("Error from building: {}", e);
-            assert_eq!(1, 2);
+    // Creates an empty directory in the current working directory
+    let dir = Builder::new().tempdir_in("").unwrap();
+    let sample_repo = "https://github.com/rust-lang/rustfmt.git";
+    let dest_path = dir.path();
+    let result = clone_git_repo(sample_repo, dest_path);
+    assert_eq!(result.is_ok(), true);
```
Comment (ytmimi) [inline check_diff/tests/git.rs:13]: Simpler to use `assert!` here since you're testing a boolean value.
```suggestion
    assert!(result.is_ok());
```

File: check_diff/tests/bash_commands.rs [hunk 17]
```diff
@@ -0,0 +1,12 @@
+use check_diff::change_directory_to_path;
+use std::env;
+use tempfile::Builder;
+
+#[test]
+fn cd_test() {
+    // Creates an empty directory in the current working directory
+    let dir = Builder::new().tempdir_in("").unwrap();
+    let dest_path = dir.path();
+    let _ = change_directory_to_path(dest_path).unwrap();
```
Comment (ytmimi) [inline check_diff/tests/bash_commands.rs:10]: `change_directory_to_path` returns `io::Result<()>`. No need to bind this to anything if you're just unwrapping:
```suggestion
    change_directory_to_path(dest_path).unwrap();
```

File: check_diff/tests/git.rs [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+fn clone_repo_test() {
+    // Creates an empty directory in the current working directory
+    let dir = Builder::new().tempdir_in("").unwrap();
+    let sample_repo = "https://github.com/rust-lang/rustfmt.git";
+    let dest_path = dir.path();
+    let result = clone_git_repo(sample_repo, dest_path);
+    assert_eq!(result.is_ok(), true);
+    let directory = fs::read_dir(dest_path);
+    // check that the directory is non-empty
+    assert_eq!(directory.iter().next().is_none(), false);
```
Comment (ytmimi) [inline check_diff/tests/git.rs:16]: Probably better to do a positive assertion. Also, you can use `.is_some()` along with `assert!` to check the boolean.

```suggestion
    let directory = fs::read_dir(dest_path);
    // check that the directory is non-empty
    assert!(directory.iter().next().is_some());
```

File: check_diff/tests/git.rs [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+    // check whether a .git folder exists in directory
+    let mut is_git_present = false;
+    for path in directory.unwrap() {
+        if path.unwrap().file_name() == ".git" {
+            is_git_present = true;
+        }
+    }
+    assert_eq!(is_git_present, true);
```
Comment (ytmimi) [inline check_diff/tests/git.rs:24]: Instead of looping we can test this more directly with [`PathBuf::exists`](https://doc.rust-lang.org/std/path/struct.Path.html#method.exists):

```suggestion
    // check whether a .git folder exists after cloning the repo
    let git_repo = dest_path.join(".git");
    assert!(git_repo.exists());
```

File: check_diff/tests/git.rs (hunk 18 above)
Comment (ytmimi) [reply check_diff/tests/git.rs:16]: This check is a little unnecessary since we're about to check that a `.git` folder exists, which would also prove that the directory isn't empty.

**Review by ytmimi — CHANGES_REQUESTED** [review]
Thanks for continuing to work on this. Take a look at my suggestions for simplifying the assertions.

**Comment by ytmimi** [comment]
From what I can tell I don't think we're actually running the `check-diff` test suite in our CI. We'll need to update [build_and_test.bat](https://github.com/rust-lang/rustfmt/blob/master/ci/build_and_test.bat) and [build_and_test.sh](https://github.com/rust-lang/rustfmt/blob/master/ci/build_and_test.sh) to cd into the `check_diff` directory and run the tests after we've run the config_proc_macro tests.

**Comment by benluiwj** [comment]
I will update the respective files - doing something similar to `config_proc_macro`

File: ci/build_and_test.bat [hunk 20]
```diff
@@ -14,6 +14,11 @@ if "%CFG_RELEASE_CHANNEL%"=="nightly" (
...
-cargo build --locked || exit /b 1
-cargo test || exit /b 1
+set crates=config_proc_macro check_diff
+(for %%crate in (%crates%)) do
+    cd %%crate || exit /b 1
+    cargo build --locked || exit /b 1
+    cargo test || exit /b 1
+    :: change back to root
+    cd .. || exit /b 1
+done
```
Comment (ytmimi) [inline ci/build_and_test.bat:24]: CI is failing. I think it might honestly just be simpler in this case to write:

```bat
:: Build and test config_proc_macro
cd config_proc_macro || exit /b 1
cargo build --locked || exit /b 1
cargo test || exit /b 1

:: Build and test check_diff
cd ..
cd check_diff || exit /b 1
cargo build --locked || exit /b 1
cargo test || exit /b 1
```

File: ci/build_and_test.sh [hunk 21]
```diff
@@ -18,6 +18,12 @@ fi
...
-cargo test
+crates=(config_proc_macro check_diff)
+for crate in "${crates[@]}"
+do
+    cd $crate
+    cargo build --locked
+    cargo test
+    # change back to root
+    cd ..
+done
```
Comment (ytmimi) [inline ci/build_and_test.sh:29]: To keep things consistent between both scripts lets revert this and apply the suggested change from https://github.com/rust-lang/rustfmt/pull/6187/files#r1678741100 but with bash syntax.

**Comment by ytmimi** [comment]
@benluiwj thanks for the updates. It's good to see CI passing. This PR has gone through several rounds of review and there are a lot of intermedia commits that could be squashed down. I think it might make sense to squash the implementation and unit tests commits into a single commit and then have another commit that updates CI. let me know if that's something you'd want to do, otherwise I can make that change.

**Review by ytmimi — APPROVED** [review]
Things are looking here!
